### PR TITLE
User linking

### LINKED ISSUE #26: User linking
Implement features around user linking:
- https://firebase.google.com/docs/reference/js/firebase.auth.Auth#fetchsigninmethodsforemail
- https://firebase.google.com/docs/reference/js/firebase.User#linkwithpopup
- https://firebase.google.com/docs/reference/js/firebase.User#linkwithredirect
- https://firebase.google.com/docs/reference/js/firebase.User#unlink

### PR DESCRIPTION
**Pull request by gustavohenke**
Closes #26.

Builds on top of #25.

### CODE REVIEW

File: auth/auth.ts [hunk 1]
```diff
@@ -43,12 +43,9 @@ export class MockAuth implements firebase.auth.Auth {
     return this.signIn(user, { isNewUser: true });
   }
 
-  fetchProvidersForEmail(email: string): Promise<any> {
```
Comment (gustavohenke) [inline auth/auth.ts:46]: likely a relic from Firebase v4 or something.

**Comment by wSedlacek** [comment]
I'll give this a review on Friday. I'm traveling right now 😅

I look forward to trying it out when I'm back.

**Comment by gustavohenke** [comment]
> I'm traveling right now

Bold! 😷

File: auth/auth.ts [hunk 2]
```diff
@@ -43,12 +43,9 @@ export class MockAuth implements firebase.auth.Auth {
...
-    throw new Error("Method not implemented.");
-  }
-
   fetchSignInMethodsForEmail(email: string): Promise<string[]> {
-    throw new Error("Method not implemented.");
+    const providers = this.store.findByEmail(email)?.providerData || [];
```
Comment (wSedlacek) [inline auth/auth.ts:47]: Having `findByEmail()` return undefined seems like a possible places that we could clean up.
If all other usages I see of `findByEmail()` we are throwing errors, maybe `findByEmail()` should just throw an error itself?
Does `fetchSignInMethodsForEmail()` in Firebase throw when you give it an invalid email?

Either way I am not a big fan of optional chaining directly after a function call.

File: auth/auth.test.ts [hunk 3]
```diff
@@ -34,6 +34,22 @@ describe("#createUserWithEmailAndPassword()", () => {
...
+    const provider = new firebase.auth.FacebookAuthProvider();
+    auth.mockSocialSignIn(provider).respondWithUser("John", "john@doe.com");
+    auth.signInWithPopup(provider);
+
+    return expect(auth.fetchSignInMethodsForEmail("foo@bar.com")).resolves.toEqual([
+      firebase.auth.EmailAuthProvider.PROVIDER_ID,
+    ]);
+  });
+});
+
```
Comment (wSedlacek) [inline auth/auth.test.ts:52]: Should we add in test for cases where the email has not been configured?

**Pull request merged**